### ISSUE TITLE
Cold Star Problem  New User

### ISSUE THREAD

**Issue by 2mrdaya** [lyst/lightfm#344]
I want to recommend items to user 4 with user feature ({"user_id":"4","user_city":"Z"}) who is not in training set. It's a cold star problem.
 I have read similar question and answer but it is not clear as you said that set all ID 0s. which IDs ?  and Where?
Please suggest what changes should i need to do in the given code and data

import numpy as np
import json
from itertools import islice
          
news_features = [{"news_id":"1","car_model":"A"},
                 {"news_id":"2","car_model":"A"},
                 {"news_id":"3","car_model":"B"},
                 {"news_id":"4","car_model":"C"}]

user_features = [{"user_id":"1","user_city":"X"},
                 {"user_id":"2","user_city":"Y"},
                 {"user_id":"3","user_city":"Z"}]

ratings = [{"user_id":"1", "news_id":"1", "rating":"1"},
           {"user_id":"1", "news_id":"2", "rating":"5"},
           {"user_id":"1", "news_id":"3", "rating":"5"},
           {"user_id":"1", "news_id":"4", "rating":"1"},
           {"user_id":"2", "news_id":"3", "rating":"4"},
           {"user_id":"2", "news_id":"1", "rating":"5"},
           {"user_id":"3", "news_id":"1", "rating":"5"},
           {"user_id":"3", "news_id":"3", "rating":"1"}]

from lightfm.data import Dataset
dataset = Dataset()

dataset.fit(users=(x['user_id'] for x in user_features),
            items=(x['news_id'] for x in news_features), 
            item_features=(x['car_model'] for x in news_features), 
            user_features=(x['user_city'] for x in user_features))

num_users, num_items = dataset.interactions_shape()

(interactions, weights) = dataset.build_interactions(((x['user_id'], x['news_id'], int(x['rating']))
                                                      for x in ratings))

item_features = dataset.build_item_features(((x['news_id'], [x['car_model']])
                                              for x in news_features))

user_features = dataset.build_user_features(((x['user_id'], [x['user_city']])
                                              for x in user_features))

from lightfm import LightFM
model = LightFM(loss='warp')
model.fit(interactions, item_features=item_features, user_features=user_features, epochs=10, num_threads=4)

### Recommended items to user 3
user_features_test = dataset.build_user_features(((x['user_id'], [x['user_city']])
                                              for x in [{"user_id":"3","user_city":"Z"}]))
scores = model.predict(2, np.arange(num_items), user_features=user_features_test)

**Comment by maciejkula** [comment]
User/item indices are indices into rows of the user/item features matrices. Construct a row in the features matrix that corresponds to the user/item you want (at some index X), then use id X to refer to that row.

Is there a reason why you ignore the code formatting I specifically ask for in the issue template?

**Comment by 2mrdaya** [comment]
Hi@maciejkula, 
Thanks for your reply.
I have just copied the code from my notebook. My problem is slightly different. Actually there is a new user 4 with user feature user_city='Z'. this user is not present in the entire dataset (not in interaction and not in the feature dataset).  What values should i pass in the given lines of codes
 at the place of "**?**" mark to recommend news. Should i need to write a function to find the user with similar feature first? 

user_features_test = dataset.build_user_features(((x['user_id'], [x['user_city']]) for x in [{"user_id":"**?**","user_city":"Z"}]))

scores = model.predict(**?**, np.arange(num_items), user_features=user_features_test)`

Can you provide a simple example with simple small dataset similar like my dataset. that can explain user/item cold star problem seperately?

**Comment by williamcao-01** [comment]
can someone help with this question? I am having the same problem too, what happens if the user is new. I tried build the user feature with the new ID, it gives the error `Number of user feature rows does not equal the number of users` @maciejkula

**Comment by V-Sher** [comment]
For anyone still looking for the answer, check out https://towardsdatascience.com/how-i-would-explain-building-lightfm-hybrid-recommenders-to-a-5-year-old-b6ee18571309
